components/scream/src/physics/p3/p3_functions_dsd2_impl.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef P3_FUNCTIONS_DSD2_IMPL_HPP
+#define P3_FUNCTIONS_DSD2_IMPL_HPP
+
+#include <fstream>
+
+#include "p3_functions.hpp"
+#include "p3_constants.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 dsd2 functions. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+void Functions<S,D>::
+get_cloud_dsd2(const Smask& qc_gt_small, const Spack& qc, Spack& nc, Spack& mu_c, const Spack& rho, Spack& nu,
+               const view_1d<const Scalar>& dnu, Spack& lamc, Spack& cdist, Spack& cdist1, const Spack& lcldm)
+{
+  lamc =   0;
+  cdist =  0;
+  cdist1 = 0;
+  nu     = 0;
+  mu_c   = 0;
+
+  if (qc_gt_small.any()) {
+    constexpr Scalar nsmall = C::NSMALL;
+    constexpr Scalar cons1  = C::CONS1;
+    // set minimum nc to prevent floating point error
+    {
+      Spack mu_c_local;
+      nc.set(qc_gt_small, pack::max(nc, nsmall));
+      mu_c_local = sp(0.0005714)*(nc * sp(1.e-6) * rho) + sp(0.2714);
+      mu_c_local = 1/(mu_c_local * mu_c_local) - 1;
+      mu_c_local = pack::max(mu_c_local, 2);
+      mu_c_local = pack::min(mu_c_local, 15);
+
+      mu_c.set(qc_gt_small, mu_c_local);
+    }
+
+    // interpolate for mass distribution spectral shape parameter (for SB warm processes)
+    if (P3C::iparam == 1) {
+      IntSmallPack dumi = IntSmallPack(mu_c) - 1;
+      Spack dnu0, dnu1;
+      pack::index_and_shift<1>(dnu, dumi, dnu0, dnu1);
+      nu.set(qc_gt_small, dnu0 + (dnu1 - dnu0) * (mu_c - Spack(dumi) - 1));
+    }
+
+    // calculate lamc
+    lamc.set(qc_gt_small, pack::cbrt(cons1 * nc * (mu_c + 3) * (mu_c + 2) * (mu_c + 1) / qc));
+
+    // apply lambda limiters
+    Spack lammin = (mu_c + 1)*sp(2.5e+4); // min: 40 micron mean diameter
+    Spack lammax = (mu_c + 1)*sp(1.e+6);   // max:  1 micron mean diameter
+
+    Smask lamc_lt_min = lamc < lammin && qc_gt_small;
```

Comment:
I suggest more use of `const auto`, but you don't need to change it here. Just a reminder.